staging/cse/windows/nvidiagpudriverfunc.ps1

```diff
@@ -0,0 +1,337 @@
+      Write-Host "Timeout $Timeout s exceeded. Stopping the installation process. Reboot for another attempt."
+      Stop-Process -InputObject $p
+    }
+    catch {
+      $Message = $_.ToString()
+      Write-Host "Exception: $Message" # the status file may get over-written when the agent re-attempts this step
+      throw
+    }
+    
+  }
+  catch [NotSupportedException] {
+    $Message = $_.ToString()
+    Write-Host $Message
+    exit 52
+  }
+  catch {
+    $FatalError += $_
+    Write-Host -ForegroundColor Red ($_ | Out-String)
+    exit $FatalError.Count
+  }
+}
+
+function Get-Setup {
+  [OutputType([hashtable])]
+    
+  # Choose driver and specific properties
+  $Driver = Select-Driver
+    
+  if ($Driver.Url -eq $null -or $Driver.CertificateUrl -eq $null) {
+    $Message = "DriverURL or DriverCertificateURL are not properly specified."
+    throw $Message
+  }
+    
+  $Setup = @{
+    RebootNeeded   = $Driver.RebootNeeded
+    CertificateUrl = $Driver.CertificateUrl
+  }
+        
+  $InstallFolder = "$PSScriptRoot\.."
+    
+  $source = $Driver.Url
+  $dest = "$InstallFolder\$($Driver.InstallExe)"
+
+  Get-DriverFile $source $dest
+
+  $Setup.Target = "$InstallFolder\$($Driver.InstallExe)"
+
+  return $Setup
+}
+
+function Select-Driver {
+  [OutputType([hashtable])]
+
+  # Set some default values
+  $Driver = @{
+    # Hard coding the FwLink for resources.json
+    ResourceUrl  = "https://go.microsoft.com/fwlink/?linkid=2181101"
+    ResourceFile = "$PSScriptRoot\..\resources.json"
+    InstallExe   = "install.exe"
+    SetupExe     = "setup.exe"
+    RebootNeeded = $false
+  }
+  $Index = @{
+    OS = 0 # Windows
+  }
+
+  # Get OS
+  # Using the following instead of VM vmetadata ( $Compute.sku -match ('.*2016-Datacenter.*|.*RS.*Pro.*') ) # 2016, Win 10
+  # to accomodate various image publishers and their SKUs
+  $OSMajorVersion = (Get-CimInstance Win32_OperatingSystem).version.split('.')[0]
+  if ( $OSMajorVersion -eq 10 ) {
+    # Win 2016, Win 10
+    $Index.OSVersion = 0
+  }
+  else {
+    # ( $OSMajorVersion -eq 6 ) Win 2012 R2
+    $Index.OSVersion = 1
+  }  
+
+  try {
+    $Compute = Get-VmData
+    $vmSize = $Compute.vmSize
+  }
+  catch {
+    Write-Host "Failed to query the SKU information. Attempting to install the extension from customer location regardless."
+  }
+
+  # Not an AzureStack scenario
+  if ($vmSize -ne $null) {
+    if ( ($Compute.vmSize -Match "_NC") -or
+        ($Compute.vmSize -Match "_ND") ) {
+      # Override if GRID driver is desired on ND or NC VMs
+      $Index.Driver = 0 # CUDA
+    }
+    elseif ( $Compute.vmSize -Match "_NV" ) {
+      # NV or Grid on ND or Grid on NC
+      $Index.Driver = 1 # GRID
+      $Driver.RebootNeeded = $true
+    }
+    else {
+      throw [NotSupportedException] "VM type $($Compute.vmSize) is not an N-series VM. Not attempting to deploy extension! More information on troubleshooting is available at https://aka.ms/NvidiaGpuDriverWindowsExtensionTroubleshoot"
```

Comment:
Please follow the current code to define an exit code and set it instead of throwing it directly.

Reply:
same in all other lines